spec/edpm-openstack-service-architecture.md

```diff
@@ -0,0 +1,326 @@
+  or partner service on the EDPM node can be done without changing the
+  DataPlane CRDs or the dataplane-operator implementation.
+
+* The service operators encapsulates all the service specific logic. Every
+  deployment, configuration or upgrade concerns for an OpenStack component can
+  be handled within a single, cohesive operator implementation.
+
+* The `CustomServiceConfig` field in the service CRDs simplifies and unifies
+  the OpenStack service configuration for the human deployer regardless where
+  the service is being deployed (i.e. podified or EDPM).
+
+**Drawbacks**:
+
+* The mechanism to watch DataPlane CRs from service operators and auto create
+  service CRs adds implementation complexity and introduces possible
+  performance impacts that could be avoided if dataplane-operator creates the
+  service CRs.
+
+* The Ready condition of the DataPlaneRole and the DataPlaneNode can only
+  represent the phase -1 and phase 0 readiness of the EDPM node. The phase 1
+  readiness can only be collected by checking the Ready condition of every
+  service CRs related to this EDPM node.
+
+
+### Example
+
+NovaCell/nova-cell1
+```yaml
+apiVersion: nova.openstack.org/v1beta1
+kind: NovaCell
+metadata:
+    name: nova-cell1
+...
+```
+No changes in NovaCell.
+
+OpenstackDataPlaneRole/edpm-compute
+```yaml
+apiVersion: dataplane.openstack.org/v1beta1
+kind: OpenStackDataPlaneRole
+metadata:
+    name: edpm-compute
+    annotations:
+        nova.openstack.org/cell: nova-cell1
+spec:
+    dataPlane: openstack-edpm
+    nodeTemplate:
+       ansiblePort: 22
+...
+```
+The `novaTemplate` is removed from the `nodeTemplate` and a new user specified
+annotation `nova.openstack.org/cell` is added so the human deployer can
+define that this Role describe a set of EDPM nodes that are compute nodes
+connected to cell1.
+
+OpenstackDataPlaneNode/edpm-compute-0
+```yaml
+apiVersion: dataplane.openstack.org/v1beta1
+kind: OpenStackDataPlaneNode
+metadata:
+    name: edpm-compute-0
+spec:
+    ansibleHost: 192.168.122.100
+    hostName: edpm-compute-0
+    node:
+        ansibleSSHPrivateKeySecret: dataplane-ansible-ssh-private-key-secret
+    role: edpm-compute
+...
+```
+The `novaTemplate` is removed from the `node`. As this DataPlaneNode refers
+to a DataPlaneRole that has a `nova.openstack.org/cell` annotation, this
+DataPlaneNode is a compute node connected to cell1.
+
+OpenstackDataPlaneNode/edpm-compute-1
+```yaml
+apiVersion: dataplane.openstack.org/v1beta1
+kind: OpenStackDataPlaneNode
+metadata:
+    name: edpm-compute-1
+    annotations:
+        nova.openstack.org/cell: nova-cell2
+spec:
+    ansibleHost: 192.168.122.101
+    hostName: edpm-compute-1
+    node:
+        ansibleSSHPrivateKeySecret: dataplane-ansible-ssh-private-key-secret
+    role: edpm-compute
+...
+```
+This DataPlaneNode has an `nova.openstack.org/cell` annotation that overrides
+the cell annotation of its role. So this EDPM node is a compute node connected
+to cell2.
+
```

Comment:
you also may prefer to have a role span multiple cells in which case you would just not set the annotation on the role and use it on the node but this make sense.

i would probably advise that OpenStackDataPlaneRole are cell alinged but there is no hard requirement either way so its good to all the felxablity to the deployer to choose how best to model there datacenter.